### ISSUE TITLE
EOF error trying to get active workflow after posting and running workflow

### ISSUE THREAD

**Issue by tn5993** [RackHD/RackHD#85]
Here is the scenerio:

Two threads tried to post then run workflow on rackhd server
One of two node failed the actions because there is some active workflow running (this behavior is expected)
The thread, which run the workflow, then tried to get active workflow. It got EOF error.

**Comment by yyscamper** [comment]
Hi @tn5993 : I just tried polling active workflow in two processes, the polling frequency is 1 sec, I don't find your mentioned `EOF error`. If could, could you please provide your code snippet?

And do you know what's the specified `EOF error` you encountered? `End Of File`?  And what's the duration between post run workflow and get active workflow?

Personally, I think this is not a RackHD issue, RackHD doesn't care you issue API request from which thread.

**Comment by heckj** [comment]
@tn5993 could you share how you have RackHD installed and if there are any proxies or systems "in front" of the API requests? I think your team has reported EOF responses cutting off what looked like normal, full responses from requests before (when getting catalog data), but we were never able to reproduce the issue. Do you have an instance of HAProxy "in front of" the RackHD proxy? Or Nginx working as a reverse proxy? Or Apache? 

I suspect the EOF issues may be related to something like because we've so far been enable to reproduce it at all.

**Comment by meganmurawski** [comment]
We have a test that uses our code to try to run multiple workflows on multiple nodes. If we have just posted a workflow against one node in one thread and another thread attempts to make an API call very quickly (gettings nodes, for example) we tend to get an EOF on the response to that request. By putting sleeps in the logic, the EOF responses seem to go away.

**Comment by heckj** [comment]
Hey @meganmurawski is this test code sharable? Can we get it (or something like it) to use as a functional or integration test to repro and ultimatley nail down the error you're seeing?

**Comment by meganmurawski** [comment]
@heckj the test is checked in (although currently pending) here: https://github.com/cloudfoundry-incubator/bosh-rackhd-cpi-release/blob/master/src/github.com/rackhd/rackhd-cpi/cpi/create_vm_test.go#L779

We can answer any questions about running it.

**Comment by heckj** [comment]
How about let's start with how do I compile and run just that test... I glanced through your wiki and readme's, but didn't see much about how to compile and run just that test, and I'm not deep on Golang to know what's probably obvious to you... a quick howto on the dev side so that we could use that test against a local instance to do some debugging would be great.

**Comment by meganmurawski** [comment]
Good point :) we can update the wiki to include some testing info. I'll get back to you on that.

**Comment by meganmurawski** [comment]
Updated the wiki to include testing and how to run a specific test using `FIt`: https://github.com/cloudfoundry-incubator/bosh-rackhd-cpi-release/wiki/Get-Started-Developing

To cause the problem more consistently, you may need to comment out these 3 lines before running the test: https://github.com/cloudfoundry-incubator/bosh-rackhd-cpi-release/blob/master/src/github.com/rackhd/rackhd-cpi/cpi/create_vm.go#L166-L168

**Comment by heckj** [comment]
reproduced the test failure, but needed to update the test code to enable my scenario. Pending PR for that at https://github.com/cloudfoundry-incubator/bosh-rackhd-cpi-release/pull/1

The test sequence is specific to having a VM created a priori for the reservation tests, so I'm not sure that this will ultimately be fruitful for replicating the tests into some usable form for reproducing, but thought the PR was worth the effort. 

```
• [SLOW TEST:7.710 seconds]
The VM Creation Workflow retrying node reservation returns an error if selection continually fails
/Users/heckj/src/bosh-rackhd-cpi-release/src/github.com/rackhd/rackhd-cpi/cpi/create_vm_test.go:628
------------------------------
•
------------------------------
• Failure [0.036 seconds]
The VM Creation Workflow retrying node reservation [It] cleans up reservation flag after receive timeout error on reserve function
/Users/heckj/src/bosh-rackhd-cpi-release/src/github.com/rackhd/rackhd-cpi/cpi/create_vm_test.go:722

  Expected error:
      <*errors.errorString | 0xc8201ff000>: {
          s: "all nodes have been reserved",
      }
      all nodes have been reserved
  not to have occurred

  /Users/heckj/src/bosh-rackhd-cpi-release/src/github.com/rackhd/rackhd-cpi/cpi/create_vm_test.go:698
------------------------------
••P•••••••••••••
```

**Comment by heckj** [comment]
@meganmurawski I think I managed to get this farther, understanding what your code is trying to do in https://github.com/cloudfoundry-incubator/bosh-rackhd-cpi-release/blob/master/src/github.com/rackhd/rackhd-cpi/cpi/create_vm_test.go#L781-L814, could you help me vet this understanding:

Your system is attempt to run the functions tryReservation 3 times in parallel, no serialization, which is calling a select function, and then a reserve function - the select function being `SelectNodeFromRackHD` and the reserve function being `reserveNodeFromRackHD`, the later of which corresponds to https://github.com/cloudfoundry-incubator/bosh-rackhd-cpi-release/blob/master/src/github.com/rackhd/rackhd-cpi/cpi/create_vm.go#L182-L195. This function in internally tries to run a workflow, leveraging RackHD's general mechanism to only allow a single workflow to be operational against a node at any given time, hence making a sort of light "reservation" against a specific node.

What I haven't been able to pin down is where you're seeing the EOF response. I've not done tracing on Go tests (or code) previously - any suggestions on how to invoke just a test on that one section that works this area of the code in a manner that would make it clear which function is failing to help nail this down?

I got some help on the slack channel and put a focus on the one file, then running `ginkgo -r .` runs just the one spec, but it appears to simply hang and there's no logging output obviously available. I'm afraid I'm completely new to even using Ginkgo for testing. It appears to just hang here:

```
 heckj  usxxheckjm1  ~  src  …  rackhd  rackhd-cpi  cpi  host_and_port +  $  ginkgo -r .
Running Suite: CPI Suite
========================
Random Seed: 1454969245
Will run 1 of 62 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
```

Any cluestick for the newb? I'm sure I'm invoking something incorrectly - suggestions?

**Comment by heckj** [comment]
from Slack chat in #bosh-rackhd-cpi channel today:

oskoss [1:51 PM] 
> I believe the EOF problem is with our golang paring the JSON recieved

[1:52] 
> Flow of control goes like this:  -> We send a request to RackHD server

[1:53] 
> Error with the `E 2016-02-18T21:49:54.311Z [on-http] [Http.Services.RestApi] [Server] Could not find graphobject with identifier library` seen in the on-http.log

[1:54] 
> we receive a response from the rackHD server that had a improperly formed JSON.

[1:54] 
> We then parse the JSON and run into an EOF error that gets thrown from GOLANG. (I.E. not RackHD)

[1:55] 
> I will do some more debugging, but I believe the issue Thinh opened probably comes from golang and can be closed. But another issue should be opened for the improperly formed JSON being returned. (basically the same issue)

heckj [2:11 PM] 
> If you want to even update the issue with those details, happy to have it - we shouldn’t be sending back improperly formed JSON, or it may be an issue that the return value is a `500` level error, and what you’re getting back is a stack trace that you’re trying to interpret at JSON, which would have the same effect.
[2:11] 
> @oskoss: ^

[2:11] 
> We do hand back tracebacks right now even over the API - may not be sane, but it was a starting point for the development...

xhe [2:24 PM] 
> Ahh good will check that out :simple_smile:
